cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/adapter/AdapterService.java

```diff
@@ -0,0 +1,509 @@
+      if (existingSpec != null) {
+        stopPrograms(namespaceId, appSpec, adapterTypeInfo, existingSpec);
+      }
+
+      startPrograms(namespaceId, appSpec, adapterTypeInfo, adapterSpec);
+      store.addAdapter(Id.Namespace.from(namespaceId), adapterSpec);
+
+    } catch (Exception e) {
+      throw new RuntimeException("Error deploying adapter application " + e.getMessage());
+    }
+  }
+
+  /**
+   * Remove adapter identified by the namespace and name.
+   *
+   * @param namespace namespace id
+   * @param adapterName adapter name
+   * @throws AdapterNotFoundException if the adapter to be removed is not found.
+   */
+  public void removeAdapter(String namespace, String adapterName) throws AdapterNotFoundException {
+    Id.Namespace namespaceId = Id.Namespace.from(namespace);
+    AdapterSpecification adapterSpec = store.getAdapter(namespaceId, adapterName);
+    if (adapterSpec == null) {
+      throw new AdapterNotFoundException(String.format("Adapter %s not found.", adapterName));
+    }
+
+    ApplicationSpecification appSpec = store.getApplication(Id.Application.from(namespaceId, adapterSpec.getType()));
+    stopPrograms(namespace, appSpec, adapterTypeInfos.get(adapterSpec.getType()), adapterSpec);
+    store.removeAdapter(namespaceId, adapterName);
+
+    // TODO: Delete the application if this is the last adapter
+  }
+
+  public void stopAdapter(String namespace, String adapterName) {
+    //TODO:
+  }
+
+  public void startAdapter(String namespace, String adapterName) {
+    //TODO:
+  }
+
+  // Deploys adapter application if it is not already deployed.
+  private ApplicationSpecification deployApplication(String namespaceId, AdapterTypeInfo adapterTypeInfo)
+    throws Exception {
+
+    ApplicationSpecification spec = store.getApplication(Id.Application.from(namespaceId, adapterTypeInfo.getType()));
+    // Application is already deployed.
+    if (spec != null) {
+      return spec;
+    }
+
+    Manager<DeploymentInfo, ApplicationWithPrograms> manager = managerFactory.create(new ProgramTerminator() {
+      @Override
+      public void stop(Id.Namespace id, Id.Program programId, ProgramType type) throws ExecutionException {
+        // no-op
+      }
+    });
+
+    Location destination = locationFactory.create(archiveDir).append(namespaceId).append(adapterTypeInfo.getType());
+    DeploymentInfo deploymentInfo = new DeploymentInfo(adapterTypeInfo.getFile(), destination);
+    ApplicationWithPrograms applicationWithPrograms = manager.deploy(Id.Namespace.from(namespaceId),
+                                                                     adapterTypeInfo.getType(), deploymentInfo).get();
+    return applicationWithPrograms.getSpecification();
+  }
+
+  // Start all the programs needed for the adapter. Currently, only scheduling of workflow is supported.
+  private void startPrograms(String namespaceId, ApplicationSpecification spec, AdapterTypeInfo adapterTypeInfo,
```

Comment:
It's not exactly starting program right? Better call it addSchedules or schedule